starlette/middleware/http.py

```diff
@@ -0,0 +1,103 @@
+from typing import AsyncGenerator, Callable, Optional, Union
+
+from .._compat import aclosing
+from ..datastructures import MutableHeaders
+from ..requests import HTTPConnection
+from ..responses import Response
+from ..types import ASGIApp, Message, Receive, Scope, Send
+
+_DispatchFlow = Union[
+    # Default case:
+    # response = yield
+    AsyncGenerator[None, Response],
+    # Early response and/or error handling:
+    # if condition:
+    #     yield Response(...)
+    #     return
+    # try:
+    #     response = yield None
+    # except Exception:
+    #     yield Response(...)
+    # else:
+    #    ...
+    AsyncGenerator[Optional[Response], Response],
+]
+
+
+class HTTPMiddleware:
+    def __init__(
+        self,
+        app: ASGIApp,
+        dispatch: Optional[Callable[[HTTPConnection], _DispatchFlow]] = None,
+    ) -> None:
+        if dispatch is None:
+            dispatch = self.dispatch
+
+        self.app = app
+        self._dispatch_func = dispatch
+
+    def dispatch(self, conn: HTTPConnection) -> _DispatchFlow:
+        raise NotImplementedError  # pragma: no cover
```

Comment:
Do we maybe want to call this `_dispatch`? I played around with the idea. My reasoning is that the main advantage of subclassing is being able to "ship" a self contained middleware, e.g. in a library. And in that case, you'd want to reduce the public API as much as possible and this way we are forcing a public method onto the class.

Reply:
I think I get your point. But not in favor of making `_dispatch()` private -- as that's public API as far as `HTTPMiddleware` is concerned. It feels odd to be required to override a private method.

Library authors who do really want to hide a dispatch implementation from their users could always do this:

```python
class MyScopedMiddleware:
    def __init__(self, app):
        self._app = HTTPMiddleware(app, dispatch=self._dispatch)

    async def _dispatch(self, __conn):
        ...

    async def __call__(self, scope, receive, send):
        await self._app(scope, receive, send)
```

Reply:
Yep reasonable, resolving this thread